runtime/storage-proxy.js

```diff
@@ -255,43 +262,76 @@ class CollectionProxy extends StorageProxyBase {
   }
 }
 
+// Variables are synchronized in a 'last-writer-wins' scheme. When the
+// VariableProxy mutates the model, it sets a barrier and expects to 
+// receive the barrier value echoed back in a subsequent update event.
+// Between those two points in time updates are not applied or
+// notified about as these reflect concurrent writes that did not 'win'.
 class VariableProxy extends StorageProxyBase {
   constructor(...args) {
     super(...args);
     this._model = null;
+    this._barrier = null;
   }
   _synchronizeModel(model) {
     assert('data' in model);
     this._model = model.data;
   }
   _updateModel(update) {
     assert('data' in update);
+    // If we have set a barrier, suppress updates until after
+    // we have seen the barrier return via an update.
+    if (this._barrier && update.barrier != this._barrier) {
+      return false;
+    }
     this._model = update.data;
+    if (this._barrier) {
+      this._barrier = null;
+      return false;
+    }
+    return true;
   }
   // Read ops: if we're synchronized we can just return the local copy of the data.
   // Otherwise, send a request to the backing store.
   // TODO: in synchronized mode, these should integrate with SynchronizeProxy rather than
   //       sending a parallel request
   get(particleId) {
     if (this._synchronized == SyncState.full) {
-      return new Promise((resolve, reject) => resolve(this._model));
+      return Promise.resolve(this._model);
     } else {
       return new Promise((resolve, reject) =>
-        this._port.HandleGet({callback: r => resolve(r), handle: this, particleId}));
+        this._port.HandleGet({callback: resolve, handle: this, particleId}));
     }
   }
-  // Write ops: in synchronized mode, any write operation will desynchronize the proxy, so
-  // subsequent reads will call to the backing store until resync is established via the update
-  // event triggered by the write.
-  // TODO: handle concurrent writes from other parties to the backing store
   set(entity, particleId) {
-    this._port.HandleSet({data: entity, handle: this, particleId});
-    this._synchronized = SyncState.pending;
+    if (JSON.stringify(this._model) == JSON.stringify(entity)) {
+      return;
+    }
+    let barrier = this._pec.generateID();
+    // TODO: is this already a clone?
+    this._model = JSON.parse(JSON.stringify(entity));
```

Comment:
Nit: only stringify entity once?

Reply:
I'm assuming we're going to clean up all of the parse/stringify at some point, eg. add a utility to compare entities for deep equality without stringification.